File: hphp/runtime/ext/fileinfo/ext_fileinfo.cpp
/*
   +----------------------------------------------------------------------+
   | HipHop for PHP                                                       |
   +----------------------------------------------------------------------+
   | Copyright (c) 2010-present Facebook, Inc. (http://www.facebook.com)  |
   | Copyright (c) 1997-2010 The PHP Group                                |
   +----------------------------------------------------------------------+
   | This source file is subject to version 3.01 of the PHP license,      |
   | that is bundled with this package in the file LICENSE, and is        |
   | available through the world-wide-web at the following url:           |
   | http://www.php.net/license/3_01.txt                                  |
   | If you did not receive a copy of the PHP license and are unable to   |
   | obtain it through the world-wide-web, please send a note to          |
   | license@php.net so we can mail you a copy immediately.               |
   +----------------------------------------------------------------------+
*/
#include <sys/types.h>
#include <sys/stat.h>

#include "hphp/runtime/ext/extension.h"
#include "hphp/runtime/base/stream-wrapper-registry.h"
#include "hphp/runtime/ext/fileinfo/libmagic/magic.h"

#include <folly/portability/Unistd.h>

namespace HPHP {
const StaticString s_finfo("finfo");

struct FileinfoResource : SweepableResourceData {
  DECLARE_RESOURCE_ALLOCATION(FileinfoResource)
  CLASSNAME_IS("file_info")
  const String& o_getClassNameHook() const override { return classnameof(); }

  explicit FileinfoResource(struct magic_set* magic) : m_magic(magic) {}
  ~FileinfoResource() override { close(); }
  void close() {
    magic_close(m_magic);
    m_magic = nullptr;
  }
  struct magic_set* getMagic() { return m_magic; }

private:
  struct magic_set* m_magic;
};

void FileinfoResource::sweep() {
  close();
}

static Variant HHVM_FUNCTION(finfo_open,
    int64_t options,
    const Variant& magic_file) {
  auto magic = magic_open(options);
  if (magic == nullptr) {
    raise_warning("Invalid mode '%" PRId64 "'.", options);
    return false;
  }

  String mf;
  if (!magic_file.isNull()) {
    mf = magic_file.toString();
  }

  auto fn = mf.empty() ? nullptr : mf.data();
  if (magic_load(magic, fn) == -1) {
    raise_warning("Failed to load magic database at '%s'.", fn);
    magic_close(magic);
    return false;
  }

  return Variant(req::make<FileinfoResource>(magic));
}

static bool HHVM_FUNCTION(finfo_close, const OptResource& finfo) {
  cast<FileinfoResource>(finfo)->close();
  return true;
}

static bool HHVM_FUNCTION(finfo_set_flags, const OptResource& finfo, int64_t options) {
  auto magic = cast<FileinfoResource>(finfo)->getMagic();
  if (magic_setflags(magic, options) == -1) {
    raise_warning(
      "Failed to set option '%" PRId64 "' %d:%s",
      options,
      magic_errno(magic),
      magic_error(magic)
    );
    return false;
  }
  return true;
}

#define FILEINFO_MODE_BUFFER 0
#define FILEINFO_MODE_STREAM 1
#define FILEINFO_MODE_FILE 2

static Variant
php_finfo_get_type(const OptResource& object, const Variant& what, int64_t options,
                   const Variant& /*context*/, int mode, int mimetype_emu) {
  String ret_val;
  String buffer;
  char mime_directory[] = "directory";
  struct magic_set *magic = NULL;

  if (mimetype_emu) {
    if (what.isString()) {
      buffer = what.toString();
      mode = FILEINFO_MODE_FILE;
    } else if (what.isResource()) {
      mode = FILEINFO_MODE_STREAM;
    } else {
      raise_warning("Can only process string or stream arguments");
    }

    magic = magic_open(MAGIC_MIME_TYPE);
    if (magic_load(magic, NULL) == -1) {
      raise_warning("Failed to load magic database.");
      goto common;
    }
  } else if (object) {
    buffer = what.toString();
    magic = cast<FileinfoResource>(object)->getMagic();
  } else {
    // if we want to support finfo as a resource as well, do it here
    not_reached();
  }

  // Set options for the current file/buffer.
  if (options) {
    HHVM_FN(finfo_set_flags)(object, options);
  }

  switch (mode) {
    case FILEINFO_MODE_BUFFER:
    {
      ret_val = magic_buffer(magic, buffer.data(), buffer.size());
      break;
    }

    case FILEINFO_MODE_STREAM:
    {
      auto stream = cast<File>(what);
      if (!stream) {
        goto common;
      }

      auto streampos = stream->tell(); // remember stream position
      stream->seek(0, SEEK_SET);

      ret_val = magic_stream(magic, stream.get());

      stream->seek(streampos, SEEK_SET);
      break;
    }

    case FILEINFO_MODE_FILE:
    {
      if (buffer.empty()) {
        raise_warning("Empty filename or path");
        ret_val.reset();
        goto clean;
      }

      auto stream = File::Open(buffer, "rb");
      if (!stream) {
        ret_val.reset();
        goto clean;
      }

      struct stat st;
      if (stream->stat(&st)) {
        if (st.st_mode & S_IFDIR) {
          ret_val = mime_directory;
        } else {
          ret_val = magic_stream(magic, stream.get());
        }
      }
      break;
    }

    default:
      raise_warning("Can only process string or stream arguments");
  }

common:
  if (!ret_val) {
    raise_warning(
      "Failed identify data %d:%s",
      magic_errno(magic),
      magic_error(magic)
    );
  }

clean:
  if (mimetype_emu) {
    magic_close(magic);
  }

  // Restore options
  if (options) {
    HHVM_FN(finfo_set_flags)(object, options);
  }

  if (ret_val) {
    return ret_val;
  }
  return false;
}

static String HHVM_FUNCTION(finfo_buffer,
    const OptResource& finfo, const Variant& string,
    int64_t options, const Variant& context) {

  String s;
  if (!string.isNull()) {
    s = string.toString();
  }
  return php_finfo_get_type(
      finfo, s, options, context,
      FILEINFO_MODE_BUFFER, 0).toString();
}

static String HHVM_FUNCTION(finfo_file,
    const OptResource& finfo, const Variant& file_name,
    int64_t options, const Variant& context) {

  String fn;
  if (!file_name.isNull()) {
    fn = file_name.toString();
  }
  return php_finfo_get_type(
      finfo, fn, options, context,
      FILEINFO_MODE_FILE, 0).toString();
}

static String HHVM_FUNCTION(mime_content_type, const Variant& filename) {
  return php_finfo_get_type(
    OptResource{}, filename, 0, uninit_null(), -1, 1
  ).toString();
}

//////////////////////////////////////////////////////////////////////////////

struct fileinfoExtension final : Extension {
  fileinfoExtension() : Extension("fileinfo", "1.0.5-dev", NO_ONCALL_YET) {}
  void moduleInit() override {
    HHVM_RC_INT(FILEINFO_NONE, MAGIC_NONE);
    HHVM_RC_INT(FILEINFO_SYMLINK, MAGIC_SYMLINK);
    HHVM_RC_INT(FILEINFO_MIME, MAGIC_MIME);
    HHVM_RC_INT(FILEINFO_MIME_TYPE, MAGIC_MIME_TYPE);
    HHVM_RC_INT(FILEINFO_MIME_ENCODING,MAGIC_MIME_ENCODING);
    HHVM_RC_INT(FILEINFO_DEVICES, MAGIC_DEVICES);
    HHVM_RC_INT(FILEINFO_CONTINUE, MAGIC_CONTINUE);
    HHVM_RC_INT(FILEINFO_PRESERVE_ATIME, MAGIC_PRESERVE_ATIME);
    HHVM_RC_INT(FILEINFO_RAW, MAGIC_RAW);
    HHVM_FE(finfo_open);
    HHVM_FE(finfo_buffer);
    HHVM_FE(finfo_file);
    HHVM_FE(finfo_set_flags);
    HHVM_FE(finfo_close);
    HHVM_FE(mime_content_type);
  }
} s_fileinfo_extension;

// Uncomment for non-bundled module
//HHVM_GET_MODULE(fileinfo);

//////////////////////////////////////////////////////////////////////////////
} // namespace HPHP

File: hphp/runtime/ext/fileinfo/ext_fileinfo.php
<?hh

class finfo {

  private $resource;

  private $options;
  private $magic_file;

  /**
   * Create a new fileinfo resource
   *
   * @param int $options - One or disjunction of more Fileinfo constants.
   * @param string $magic_file - Name of a magic database file, usually
   *   something like /path/to/magic.mime. If not specified, the MAGIC
   *   environment variable is used. If the environment variable isn't set,
   *   then PHP's bundled magic database will be used.   Passing NULL or an
   *   empty string will be equivalent to the default value.
   */
  public function __construct(int $options = FILEINFO_NONE,
                              ?string $magic_file = NULL) {
    $this->resource = finfo_open($options, $magic_file);

    $this->options = $options;
    $this->magic_file = $magic_file;
  }

  public function __sleep() {
    return vec['options', 'magic_file'];
  }

  public function __wakeup() {
    $this->resource = finfo_open($this->options, $this->magic_file);
  }

  /**
   * Return information about a string buffer
   *
   * @param string $string - Content of a file to be checked.
   * @param int $options - One or disjunction of more Fileinfo constants.
   * @param resource $context -
   *
   * @return string - Returns a textual description of the string
   *   argument, or FALSE if an error occurred.
   */
  public function buffer(?string $string = NULL,
                         int $options = FILEINFO_NONE,
                         ?resource $context = NULL): string {
    return finfo_buffer($this->resource, $string, $options, $context);
  }

  /**
   * Return information about a file
   *
   * @param string $file_name - Name of a file to be checked.
   * @param int $options - One or disjunction of more Fileinfo constants.
   * @param resource $context - For a description of contexts, refer to .
   *
   * @return string - Returns a textual description of the contents of
   *   the filename argument, or FALSE if an error occurred.
   */
  public function file(?string $file_name = NULL,
                       int $options = FILEINFO_NONE,
                       ?resource $context = NULL): string {
    return finfo_file($this->resource, $file_name, $options, $context);
  }

  /**
   * Set libmagic configuration options
   *
   * @param int $options - One or disjunction of more Fileinfo constants.
   *
   * @return bool -
   */
  public function set_flags(int $options): bool {
    $ret = finfo_set_flags($this->resource, $options);
    if ($ret) {
      $this->options = $options;
    }
    return $ret;
  }

}

/**
 * Return information about a string buffer
 *
 * @param resource $finfo - Fileinfo resource returned by finfo_open().
 * @param string $string - Content of a file to be checked.
 * @param int $options - One or disjunction of more Fileinfo constants.
 * @param resource $context -
 *
 * @return string - Returns a textual description of the string argument,
 *   or FALSE if an error occurred.
 */
<<__Native>>
function finfo_buffer(resource $finfo,
                      ?string $string = NULL,
                      int $options = FILEINFO_NONE,
                      ?resource $context = NULL): string;

/**
 * Close fileinfo resource
 *
 * @param resource $finfo - Fileinfo resource returned by finfo_open().
 *
 * @return bool -
 */
<<__Native>>
function finfo_close(resource $finfo): bool;

/**
 * Return information about a file
 *
 * @param resource $finfo - Fileinfo resource returned by finfo_open().
 * @param string $file_name - Name of a file to be checked.
 * @param int $options - One or disjunction of more Fileinfo constants.
 * @param resource $context - For a description of contexts, refer to .
 *
 * @return string - Returns a textual description of the contents of the
 *   filename argument, or FALSE if an error occurred.
 */
<<__Native>>
function finfo_file(resource $finfo,
                    ?string $file_name = NULL,
                    int $options = FILEINFO_NONE,
                    ?resource $context = NULL): string;

/**
 * Create a new fileinfo resource
 *
 * @param int $options - One or disjunction of more Fileinfo constants.
 * @param string $magic_file - Name of a magic database file, usually
 *   something like /path/to/magic.mime. If not specified, the MAGIC
 *   environment variable is used. If the environment variable isn't set,
 *   then PHP's bundled magic database will be used.   Passing NULL or an
 *   empty string will be equivalent to the default value.
 *
 * @return resource - (Procedural style only) Returns a magic database
 *   resource on success.
 */
<<__Native>>
function finfo_open(int $options = FILEINFO_NONE,
                    ?string $magic_file = NULL): mixed;

/**
 * Set libmagic configuration options
 *
 * @param resource $finfo - Fileinfo resource returned by finfo_open().
 * @param int $options - One or disjunction of more Fileinfo constants.
 *
 * @return bool -
 */
<<__Native>>
function finfo_set_flags(resource $finfo,
                         int $options): bool;

/**
 * Detect MIME Content-type for a file (deprecated)
 *
 * @param string $filename - Path to the tested file.
 *
 * @return string - Returns the content type in MIME format, like
 *   text/plain or application/octet-stream.
 */
<<__Native>>
function mime_content_type(mixed $filename): string;
